lib/internal/modules/esm/translators.js

```diff
@@ -25,13 +28,27 @@ const debug = debuglog('esm');
 const translators = new SafeMap();
 module.exports = translators;
 
+function initializeImportMeta(meta, { url }) {
+  meta.url = url;
+}
+
+async function importModuleDynamically(specifier, { url }) {
+  const loader = await esmLoader.loaderPromise;
+  return loader.import(specifier, url);
+}
+
 // Strategy for loading a standard JavaScript module
 translators.set('esm', async (url) => {
   const source = `${await readFileAsync(new URL(url))}`;
   debug(`Translating StandardModule ${url}`);
+  const module = new ModuleWrap(stripShebang(source), url);
+  callbackMap.set(module, {
+    initializeImportMeta,
+    importModuleDynamically,
+  });
```

Comment:
Could we not do something here where we have a default loader behaviour that is then overridden by the callbackMap for VM?

That would then avoid the need to call this set on each and every module in the base-level loader.

Reply:
I would not like any "defaults" or "fallbacks" because it opens the door for escaping contexts.

Reply:
Can you explain how VM could "lose track" of something if we did this?

Reply:
Well I'm not sure, people find all sorts of weird ways to do things. I can however say that we elimate all that by treating all the wraps the same. Is there any specific problem with the current design you want be to look at?

Reply:
I just have a strong belief in doing minimal work, and aiming for the simplest possible implementation, especially when laying code foundations.

Reply:
i would say that because all modules interact with this flow in the same way, it is already as simple as possible.

Reply:
This is creating a code dependency of the module loader on VM, which is a first. It is also running code to set modules in a map that wouldn't be necessary if the default loader was assumed.

Every piece of code should justify its existence based on its benefit, and I'm not sure I see what the strong benefit here is to setting ALL modules in a context map, when we only need to distinguish those modules that have unique contexts.

Reply:
ModuleWrap and ContextifyScript are equally part of our VM infra so I don't really see your point here. Also this is scoped by script not context.

Reply:
> This is creating a code dependency of the module loader on VM, which is a first.

I think it's important to clarify here that the module loader doesn't depend on the _bulk_ of the VM, only a single `WeakMap` that arguably can be put anywhere else, even under `lib/internal/modules/esm/` if that's preferred. In fact, the only reason why it's a `WeakMap` in the first place is because V8 doesn't have private instance variables yet.

I also echo @devsnek's point about not treating the system loader any more special: IMO there should be no "default" behavior for a module – though I am open to be convinced otherwise.

Reply:
The base-level loader is privileged in the sense that it is the default code path for `node x`. So my argument here is that the default behaviour of `import('x')` should similarly assume the same code path, with opt-outs for overridden behaviours. I see no reason to individually bind the callbacks for dynamic import in the default loader, when it is plainly the default behaviour.

I would be open to lowering the implementation of the context below VM, as yes, its the coupling to VM that I'm largely against here.

Reply:
my design principle is that because the base-level loader is privileged, getting to interact with it is also a privilege. i don't want things messing with it unless i explicitly want them to.

Reply:
So as mentioned my two concerns are:

1. From a performance perspective we are unnecessarily tracking modules in a map, that could otherwise assume the base-level realm loader without sacrificing security.
2. We are coupling the loader to vm

(1) I can compromise on.

But (2) is what I would really like to avoid by either tackling (1) or by lowering the context code into a module other than vm.

Reply:
@guybedford i can rename the file the map is in if you want. its not really coupled to "vm" in any other way, and like i said i consider ModuleWrap part of vm.

Reply:
Certainly it could be renamed `internal/modules/esm/module_context.js` or something like that.

And the code on the C++ side that handles this callback is all in node_contextify right?

Reply:
we call into JS from ModuleWrap::ImportModuleDyanmically

Reply:
`callbackMap` is no longer in the `vm` namespace.

Reply:
Nice, thanks